uaclient/api/u/pro/auto_attach/with_short_retry/v1.py

```diff
@@ -0,0 +1,243 @@
+            # We are attached on non-Pro Image, just report already attached
+            raise exceptions.AlreadyAttachedError(cfg)
+        if isinstance(e, exceptions.CloudFactoryNoCloudError):
+            raise exceptions.UserFacingError(
+                messages.UNABLE_TO_DETERMINE_CLOUD_TYPE
+            )
+        if isinstance(e, exceptions.CloudFactoryNonViableCloudError):
+            raise exceptions.UserFacingError(messages.UNSUPPORTED_AUTO_ATTACH)
+        if isinstance(e, exceptions.CloudFactoryUnsupportedCloudError):
+            raise exceptions.NonAutoAttachImageError(
+                messages.UNSUPPORTED_AUTO_ATTACH_CLOUD_TYPE.format(
+                    cloud_type=e.cloud_type
+                )
+            )
+        # we shouldn't get here, but this is a reasonable default just in case
+        raise exceptions.UserFacingError(
+            messages.UNABLE_TO_DETERMINE_CLOUD_TYPE
+        )
+
+    if not instance:
+        # we shouldn't get here, but this is a reasonable default just in case
+        raise exceptions.UserFacingError(
+            messages.UNABLE_TO_DETERMINE_CLOUD_TYPE
+        )
+    return instance
+
+
+def detach_before_auto_attach(cfg: UAConfig):
+    current_iid = identity.get_instance_id()
+    if cfg.is_attached:
+        prev_iid = cfg.read_cache("instance-id")
+        if str(current_iid) == str(prev_iid):
+            raise exceptions.AlreadyAttachedOnPROError(str(current_iid))
+        print("Re-attaching Ubuntu Pro subscription on new instance")
+        if _detach(cfg, assume_yes=True) != 0:
+            raise exceptions.UserFacingError(
+                messages.DETACH_AUTOMATION_FAILURE
+            )
+
+
+def get_auto_attach_token(cfg: UAConfig, cloud: AutoAttachCloudInstance):
+    contract_client = contract.UAContractClient(cfg)
+    try:
+        tokenResponse = contract_client.request_auto_attach_contract_token(
+            instance=cloud
+        )
+    except exceptions.ContractAPIError as e:
+        if e.code and 400 <= e.code < 500:
+            raise exceptions.NonAutoAttachImageError(
+                messages.UNSUPPORTED_AUTO_ATTACH
+            )
+        raise e
+
+    token = tokenResponse["contractToken"]
+    return token
+
+
+def with_short_retry(options: AutoAttachWithShortRetryOptions, cfg: UAConfig):
+    i = 0
+    already_attached = False  # avoid AlreadyAttachedOnProError after 1st time
+    while i <= 2:
+        if not already_attached:
+            instance = None  # type: Optional[AutoAttachCloudInstance]
+            token = None  # Optional[str]
+            try:
+                instance = get_cloud_instance(cfg)
+            except exceptions.UserFacingError as e:
+                raise CloudError(
+                    msg=str(e.msg), msg_code="unable-to-complete-cloud-checks"
+                )
+            if not instance:
+                raise AutoAttachDisabledError(
+                    msg="UA Config disable_auto_attach is set."
+                )
```

Comment:
We don't need to convert the exceptions into `APIErrors` here. The `api` module already covers that for us. This apply to all other functions we are manually calling here